code/controlconfig/controlsconfig.cpp

```diff
@@ -1824,58 +1982,65 @@ int control_config_draw_list(int select_tease_line) {
-	int w, x, y, conflict;
-	int k; // polled key.  Can be masked with SHIFT and/or ALT
-	short j = JOY_TOTAL_BUTTONS; // polled joy button
-	short joy = -1;              // polled joystick id
-	int z = Cc_lines[Selected_line].cc_index; // Selected line's cc_index; value: (z &= ~JOY_AXIS); Is an axis index if (z & JOY_AXIS) == true;
-	int font_height = gr_get_font_height();
-	int select_tease_line = -1;  // line mouse is down on, but won't be selected until button released
-	static float timer = 0.0f;
-	static UI_TIMESTAMP bound_timestamp;
-	static char bound_string[40];
-	
-	timer += frametime;
+	bool bind = false; // is true if binding should happen.  Actually is an "Input detected" flag.
+	bool done = false; // is true if we're done binding and ready for exiting this mode
 
-	if (Binding_mode) {
-		bool bind = false;	// is true if binding should happen.  Actually is an "Input detected" flag.
-		bool done = false;	// is true if we're done binding and ready for exiting this mode
+	//API mode needs to run this once before beginning
+	if (API_Access && (Binding_mode != 1)) {
+		control_config_do_bind(API_Access);
+	}
+
+	// Poll for keypress
+	int k;
 
-		// Poll for keypress
-		k = game_poll();
+	if (!API_Access)
+		k = game_poll(); // polled key.  Can be masked with SHIFT and/or ALT
+	else {
+		// In API-Access mode, game_poll has already happened this frame. Either by the state we're actually running in,
+		// or by the game_poll hardcoded into overrided OnFrame hooks. Hence, polling again is incorrect. The key is
+		// available in Current_key_down, and mouse and joy states are still up-to-date
+		extern int Current_key_down;
+		k = Current_key_down;
+	}
+
+	if (!API_Access) {
 		Ui_window.use_hack_to_get_around_stupid_problem_flag = 1;
 		Ui_window.process(0);
+	}
 
-		if (k == KEY_ENTER) {
-			// Cancel axis bind if Enter is pressed
-			bind = true;
-		}
+	if (k == KEY_ENTER) {
+		// Cancel axis bind if Enter is pressed
+		bind = true;
+	}
 
-		// Poll for joy btn presses
-		// Stop polling all joys if a btn was detected
-		for (joy = CID_JOY0; joy < CID_JOY_MAX; joy++) {
-			if (!joy_present(joy)) {
-				continue;
-			}
-			for (j = 0; j < JOY_TOTAL_BUTTONS; j++) {
-				if (joy_down_count(CC_bind(static_cast<CID>(joy), j), 1)) {
-					// btn is down, save it in j and joy
-					// Cancel axis bind if any button is pressed
-					bind = true;
-					goto bind_find_joy_btn;
-				}
+	short j = JOY_TOTAL_BUTTONS; // polled joy button
+	short joy = -1;              // polled joystick id
+
+	// Poll for joy btn presses
+	// Stop polling all joys if a btn was detected
+	for (joy = CID_JOY0; joy < CID_JOY_MAX; joy++) {
+		if (!joy_present(joy)) {
+			continue;
+		}
+		bool brk = false;
+		for (j = 0; j < JOY_TOTAL_BUTTONS; j++) {
+			if (joy_down_count(CC_bind(static_cast<CID>(joy), j), 1)) {
+				// btn is down, save it in j and joy
+				// Cancel axis bind if any button is pressed
+				bind = true;
+				brk = true;
```

Comment:
Was there a problem using `goto` to break out of both loops?

Reply:
My impression was the goto was not the preferred way of doing things anymore so while I was mucking around with the code I changed them to breaks.

Reply:
`goto` does indeed have a bad rap which is mostly deserved, but in very rare cases it makes for cleaner code without the use of breakout variables (such as what you've done here).

I won't call for a change here, but I thought it would be a good opportunity to point it out here.